cmd/snap-update-ns/secure_bindmount.go

```diff
@@ -0,0 +1,85 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"syscall"
+)
+
+// SecureBindMount performs a bind mount between two absolute paths
+// containing no symlinks, using a private stash directory as an
+// intermediate step
+func SecureBindMount(sourceDir, targetDir string, flags uint) error {
+	// This function only attempts to handle bind mounts
+	if flags&syscall.MS_BIND == 0 {
+		return fmt.Errorf("only bind mounts are supported")
+	}
+	// The kernel doesn't support recursively switching a tree of
+	// bind mounts to read only, and we haven't written a work
+	// around.
+	if flags&syscall.MS_RDONLY != 0 && flags&syscall.MS_REC != 0 {
+		return fmt.Errorf("cannot use MS_RDONLY and MS_REC together")
+	}
+
+	// Step 1: acquire file descriptors representing the source
+	// and destination directories, ensuring no symlinks are
+	// followed.
+	sourceFd, err := secureOpenPath(sourceDir)
+	if err != nil {
+		return err
+	}
+	defer sysClose(sourceFd)
+	targetFd, err := secureOpenPath(targetDir)
+	if err != nil {
+		return err
+	}
+	defer sysClose(targetFd)
+
+	// Step 2: perform a bind mount between the paths identified
+	// by the two file descriptors.
```

Comment:
Please adjust the comment here:
```
// Step 2: perform a bind mount between the paths identified
// by the two file descriptors. We primarily care about privilege
// escalation here and trying to race the sysMount() by removing
// any part of the path to replace it with a symlink will cause
// the fd path to be marked as '(deleted)' and therefore
// unmountable. Renaming the directory causes the mount to happen
// with the newly renamed path, but the rename is controlled by
// DAC so while the user could race the mount source or target,
// this rename can't be used to gain privileged access to files.
// For systems with AppArmor enabled, this raced rename would
// be denied by the per-snap snap-update-ns AppArmor profle.
```

Reply:
@jhenstridge - I see you made one comment change (thanks!) tcan you make this one too?

Reply:
I don't think that is an accurate description of what is happening inside the kernel: as mentioned in my previous comment, I don't think the `readlink` value actually matters.

Digging in, the inode operations for these links define both `get_link` and `readlink`:

https://github.com/torvalds/linux/blob/master/fs/proc/base.c#L1674-L1678

Usually a file system will implement `get_link` to return the value of a symlink, and rely on the `generic_readlink` implementation (which just calls `get_link` and copies the result to user space memory).

Here instead, we've got a `get_link` implementation that uses `nd_jump_link` to short circuit the symlink resolution process using a provided `struct path` (which is the one associated with the file descriptor for the fd symlinks), and a custom `readlink` implementation for when user space actually wants the symlink value rather than just resolving it.

So the `(deleted)` suffix shown by readlink doesn't actually matter here: what actually makes a difference is that the dentry associated with the deleted directory has been marked with DCACHE_CANT_MOUNT.

Reply:
Well, I phrased the comment for readability but I guess I didn't pull it off. The DCACHE_CANT_MOUNT is of course what prevents the mount. The stuff about symlinks isn't about touching the entries in /proc/pid/fd, but with trying to manipulate the normal, non-proc paths after we're holding the open fd.

How about this:
```
// Step 2: perform a bind mount between the paths identified
// by the two file descriptors. We primarily care about privilege
// escalation here and trying to race the sysMount() by removing
// any part of the dir (sourceDir or targetDir) after we have an
// open file descriptor to it (sourceFd or targetFd) to then
// replace an element of the dir's path with a symlink will cause
// the fd path (ie, sourceFdPath or targetFdPath) to be marked as
// unmountable within the kernel (this path is also changed to
// show as '(deleted)'). Alternatively, simply renaming the dir
// (sourceDir or targetDir) after we have an open file descriptor
// to it (sourceFd or targetFd) causes the mount to happen with
// the newly renamed path, but this rename is controlled by DAC
// so while the user could race the mount source or target, this
// rename can't be used to gain privileged access to files. For
// systems with AppArmor enabled, this raced rename would be
// denied by the per-snap snap-update-ns AppArmor profle.
```